submissions/vovan-zt/Frogger-Game/app.js

```diff
@@ -0,0 +1,108 @@
+const gridSize = {
+    width: 101,
+    height: 81,
+};
+
+const fieldSize = {
+width: gridSize.width*5,
+height: gridSize.height*6,
+top:0,
+left:0,
+};
+
+const startPlayerPosition = {
+x:gridSize.width*2 ,
+y:gridSize.height*5,
+};
+
+const enemySpeed = 200;
+const  Enemy = function(x, y, speed) {  
+    this.x = x;
+    this.y = y;
+    this.speed = speed;
+    this.sprite = 'images/enemy-bug.png';
+};
+
+Enemy.prototype.update = function(dt) {
+    this.x += this.speed*dt;
+
+    if (this.x > fieldSize.width) {
+        this.x = -gridSize.width;
+        this.speed = 100 + Math.floor(Math.random() * enemySpeed);
+    };
+
+    if (player.x + gridSize.width > this.x && player.x < this.x + gridSize.width &&
+        player.y + gridSize.height > this.y && player.y < this.y + gridSize.height) {
+        player.x = startPlayerPosition.x;
+        player.y = startPlayerPosition.y;
+    }  
+};
+
+Enemy.prototype.render = function() {
+    ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+const  Player = function(x, y) {
+    this.x = x;
+    this.y = y;
+    this.sprite = 'images/char-boy.png';
+};
+
+Player.prototype.update = function (dt) {
+ 
+};
+
+Player.prototype.render = function () {
+    ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+Player.prototype.handleInput = function (keyPress) {
+    if (keyPress == 'up' && this.y > fieldSize.top) {
+        this.y -= gridSize.height;
+    }
+
+    if (keyPress == 'down' && this.y < fieldSize.width-gridSize.width) {
+        this.y += gridSize.height;
+    }
+
+    if (keyPress == 'left' && this.x > fieldSize.left) {
+        this.x -= gridSize.width;
+    }
+
+    if (keyPress == 'right' && this.x < fieldSize.width-gridSize.width) {
+        this.x += gridSize.width;
+    }
+
+    if (this.y <= fieldSize.top) {
+        setTimeout(() => {
+            this.x = startPlayerPosition.x;
+            this.y = startPlayerPosition.y;
+        }, 600);
+    };
+}; 
+
+let positionEnemies= [];
+function numbeOfEnemies(amount) {
+    for(let i = 1; i<=amount; i++) {
+        if (i>3) {
+        return numbeOfEnemies(amount-i+1);
+        }
+        positionEnemies.push(i);
+    }
+    positionEnemies = positionEnemies.map(rowNumber => rowNumber * gridSize.height);
+    return positionEnemies;
```

Comment:
As a peer developer I cannot grasp the idea behind this recursive call.
It seems that by the moment we have already 3 values (when `i === 4`) we just discard what's been done so far and we just create a new array of something with a different input value (`amount-i+1`). And we never go past 4.
Oh, here we've got `positionEnemies` global.... So the collected values are not discarded completely.
Very obscure fragment.
As a peer developer I cannot understand quickly what is going on here and why it is implemented in this way...
Can we please make this fragment more obvious?
Imagine, you show this project as a part of your interview and the interviewer wants to look into your code.

Reply:
I wanted to make a feature so that you can add any number of enemies. The first thing that came to mind I wrote :)

Reply:
This function doesn't create any enemies. We cannot expect it to return any enemies.
What it does so that it helps us to create any number of enemies?

Reply:
You are doing right by splitting a complex task into fragments where each does a smaller job and integral result is big.
So you just need to clearly articulate what you expect to have on every step of the way from the **number of enemies** to the **list of instantiated enemies**.

Reply:
![TEST](https://user-images.githubusercontent.com/75982068/187036315-c84e2510-cd7d-45ae-beb9-93502cba1c09.png)

It creates the array I need :)

Reply:
const positionEnemies = {
    first: gridSize.height * 1,
    second: gridSize.height * 2,
    third: gridSize.height * 3,
}

const allEnemies = [
    positionEnemies.first,
    positionEnemies.second,
    positionEnemies.third,
].map((location) => (location = new Enemy(location, location, enemySpeed)))


I can do this and remove that function

or so



const allEnemies = [
    positionEnemies.first,
    positionEnemies.second,
    positionEnemies.third,
].map((positionY) => (positionY = new Enemy(this.x, positionY, enemySpeed)))